### PR TITLE
Executors types

### PR DESCRIPTION
**Pull request by InvincibleRMC**

### CODE REVIEW

File: rclpy/rclpy/executors.py [hunk 1]
```diff
@@ -597,37 +640,40 @@ def _wait_for_ready_callbacks(
...
-                clients.extend(filter(self.can_execute, node.clients))
-                services.extend(filter(self.can_execute, node.services))
-                node_guards = filter(self.can_execute, node.guards)
-                waitables.extend(filter(self.can_execute, node.waitables))
-                # retrigger a guard condition that was triggered but not handled
-                for gc in node_guards:
-                    if gc._executor_triggered:
-                        gc.trigger()
-                    guards.append(gc)
+            if nodes_to_use:
```
Comment (sloretz) [inline rclpy/rclpy/executors.py:657]: Nit, if I follow correctly `nodes_to_use` is always a list, so we don't need to check `if nodes_to_use` because the `for` loop will do nothing in the case of an empty list.

File: rclpy/rclpy/executors.py [hunk 2]
```diff
@@ -727,53 +776,56 @@ def _wait_for_ready_callbacks(
                         gc._executor_triggered = True
 
                 # Check waitables before wait set is destroyed
+                if nodes_to_use:
```
Comment (sloretz) [inline rclpy/rclpy/executors.py:779]: Same nit about `nodes_to_use`

File: rclpy/rclpy/executors.py [hunk 3]
```diff
@@ -727,53 +776,56 @@ def _wait_for_ready_callbacks(
...
+                        for wt in node.waitables:
+                            # Only check waitables that were added to the wait set
+                            if wt in waitables and wt.is_ready(wait_set):
+                                if wt.callback_group.can_execute(wt):
+                                    handler = self._make_handler(wt, node, self._take_waitable)
+                                    yielded_work = True
+                                    yield handler, wt, node
+
+            # Process ready entities one node at a time
+            if nodes_to_use:
```
Comment (sloretz) [inline rclpy/rclpy/executors.py:790]: Same nit about `nodes_to_use`

**Review by sloretz — COMMENTED** [review]
LGTM with a couple nitpicks addressed!

File: rclpy/rclpy/executors.py [hunk 4]
```diff
@@ -387,15 +416,19 @@ def check_argument_type(callback_func, target_type):
 
                 # User might change the Timer.callback function signature at runtime,
                 # so it needs to check the signature every time.
-                arg_name = check_argument_type(tmr.callback, target_type=TimerInfo)
-                prefilled_arg = {arg_name: timer_info}
+                if tmr.callback:
+                    arg_name = check_argument_type(tmr.callback, target_type=TimerInfo)
```
Comment (fujitatomoya) [inline rclpy/rclpy/executors.py:420]: these code area is not directly related to type checking, but refactoring makes sense to me. thanks!

File: rclpy/rclpy/executors.py [hunk 5]
```diff
@@ -742,10 +790,11 @@ def _wait_for_ready_callbacks(
...
                         if tmr.handle.is_timer_ready():
-                            if tmr.callback_group.can_execute(tmr):
-                                handler = self._make_handler(tmr, node, self._take_timer)
-                                yielded_work = True
-                                yield handler, tmr, node
+                            if tmr.callback_group:
+                                if tmr.callback_group.can_execute(tmr):
+                                    handler = self._make_handler(tmr, node, self._take_timer)
+                                    yielded_work = True
+                                    yield handler, tmr, node
```
Comment (fujitatomoya) [inline rclpy/rclpy/executors.py:797]:
```suggestion
                            if tmr.callback_group and tmr.callback_group.can_execute(tmr):
                                handler = self._make_handler(tmr, node, self._take_timer)
                                yielded_work = True
                                yield handler, tmr, node
```

File: rclpy/rclpy/executors.py [hunk 6]
```diff
@@ -756,7 +805,7 @@ def _wait_for_ready_callbacks(
 
                 for gc in node.guards:
                     if gc._executor_triggered:
-                        if gc.callback_group.can_execute(gc):
+                        if gc.callback_group and gc.callback_group.can_execute(gc):
```
Comment (fujitatomoya) [inline rclpy/rclpy/executors.py:808]: Why do we not need to add the same check for waitables, subscriptions, clients and services?

File: rclpy/rclpy/executors.py [hunk 7]
```diff
@@ -844,8 +895,9 @@ def _spin_once_impl(
             pass
         else:
             handler()
-            if handler.exception() is not None:
-                raise handler.exception()
+            exception = handler.exception()
+            if exception is not None:
+                raise exception
```
Comment (fujitatomoya) [inline rclpy/rclpy/executors.py:900]: I think this was kind of bug to cause the internal state mismatch? when `handler.exception()` is called `_exception_fetched` is already true, that means we should keep that fetched exception like this. just checking.

**Review by fujitatomoya — APPROVED** [review]
overall lgtm with green CI, a few comments.

File: rclpy/rclpy/callback_groups.py [hunk 8]
```diff
@@ -24,7 +24,8 @@
     from rclpy.service import Service
     from rclpy.waitable import Waitable
     from rclpy.guard_condition import GuardCondition
-    Entity = Union[Subscription, Timer, Client, Service, Waitable[Any], GuardCondition]
+    Entity = Union[Subscription[Any], Timer, Client[Any, Any, Any], Service[Any, Any, Any],
```
[the remark this answers is not in the document]
Comment (fujitatomoya) [reply rclpy/rclpy/callback_groups.py]: Ah, you already addressed this one. ignore me, sorry.

File: rclpy/rclpy/executors.py (hunk 6 above)
Comment (InvincibleRMC) [reply rclpy/rclpy/executors.py:808]: Since `GuardCondition` can have Optional `.callback_group`. And the others always have a `CallbackGroup`

File: rclpy/rclpy/executors.py (hunk 7 above)
Comment (InvincibleRMC) [reply rclpy/rclpy/executors.py:897]: Yes I believe it could cause a mismatch in case somehow some other function got called in between so storing the exception should avoid it.

**Comment by InvincibleRMC** [comment]
@fujitatomoya I fixed a typing mistake should be good to re-run.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21822/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16092/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1324/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22522/)

**Comment by InvincibleRMC** [comment]
@fujitatomoya is this good to merge in?
